java/src/main/native/src/row_conversion.cu

```diff
@@ -0,0 +1,636 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <limits>
+
+#include <cudf/table/table.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sequence.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include "row_conversion.hpp"
+
+namespace cudf {
+namespace java {
+
+template <typename T>
+std::unique_ptr<rmm::device_uvector<T>> copy_to_dev_async(
+        const std::vector<T> & input,
+        cudaStream_t stream,
+        rmm::mr::device_memory_resource* mr) {
+    std::unique_ptr<rmm::device_uvector<T>> ret(new rmm::device_uvector<T>(
+                input.size(), stream, mr));
+    CUDA_TRY(cudaMemcpyAsync(ret->data(),
```

Comment:
My understanding of asynchronous CUDA `memcpy()` is iffy. From [the manual](https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async), I understand that this copy from non-pinned host memory to device-memory? Does any synchronization need to be done before the result of the copy is first accessed? (Or is it so fast that it doesn't really matter?)

Reply:
No synchronization is needed if the copy and the kernel that reads the data both happen on the same stream.  CUDA guarantees FIFO order within a stream.  The time we would have to worry is when we are copying data to the host, because the host does not explicitly wait or if we will use a different stream to read the data.

Reply:
I see, thanks. Here's a bug that reminded me of this conversation:
https://github.com/rapidsai/cudf/issues/6700